OpenSim/Examples/Plugins/VolumetricContactExample/VolumetricContactForce.h

```diff
@@ -0,0 +1,134 @@
+    : public TwoFrameLinker<Force, PhysicalFrame>
+{
+    OpenSim_DECLARE_CONCRETE_OBJECT(VolumetricContactForce, TwoFrameLinker);
+public:
+//=============================================================================
+// PROPERTIES
+//=============================================================================
+    // TODO: look for an osimsimulation library registered type, and attach it to that library instead of a separate one?
+
+    OpenSim_DECLARE_PROPERTY(k_V, double,
+        "Volumetric stiffness (N/m^3)");
+    OpenSim_DECLARE_PROPERTY(a_V, double,
+        "Volumetric damping");
+    OpenSim_DECLARE_PROPERTY(mu_s, double,
+        "Static friction");
+    OpenSim_DECLARE_PROPERTY(mu_d, double,
+        "Dynamic friction");
+    OpenSim_DECLARE_PROPERTY(v_t, double,
+        "Friction transition velocity (m/s)");
+    OpenSim_DECLARE_PROPERTY(w_t, double,
+        "Friction angular transition velocity (rad/s)");
+    OpenSim_DECLARE_PROPERTY(ellipsoid_dims, SimTK::Vec3,
+        "Ellipsoid dimensions (x, y, z axis) (m)");
+
+//=============================================================================
+// METHODS
+//=============================================================================
+public:
+    // Default Constructor
+    VolumetricContactForce();
+
+    VolumetricContactForce(const std::string& name,
+        const PhysicalFrame& ellipsoidFrame, const SimTK::Transform& transformInEllipsoidFrame,
+        const PhysicalFrame& planeFrame, const SimTK::Transform& transformInPlaneFrame,
+        const double& k_V,
+        const double& a_V,
+        const double& mu_s,
+        const double& mu_d,
+        const double& v_t,
+        const double& w_t,
+        const SimTK::Vec3& ellipsoid_dims);
+
+    //--------------------------------------------------------------------------
+    // COMPUTATION
+    //--------------------------------------------------------------------------
+    /** 
+      */
+    void computeForce(const SimTK::State& s, 
+                      SimTK::Vector_<SimTK::SpatialVec>& bodyForces, 
+                      SimTK::Vector& generalizedForces) const override;
+
+    /** This may be more difficult for volumetric... */
+    double computePotentialEnergy(const SimTK::State& s) const override;
+
+    //-----------------------------------------------------------------------------
+    // Reporting
+    //-----------------------------------------------------------------------------
+    /** 
+     * Provide name(s) of the quantities (column labels) of the force value(s) to be reported
+     */
+    OpenSim::Array<std::string> getRecordLabels() const override;
+    /**
+    *  Provide the value(s) to be reported that correspond to the labels
+    */
+    OpenSim::Array<double> getRecordValues(const SimTK::State& state) const override;
+
+protected:
+    // TODO: is there any reason this should be implemented? It was in BodyDragForce
+    //virtual void connectToModel(Model& aModel);
```

Comment:
`Component::connectToModel` is not virtual. It should be removed from `BodyDragForce`. If there is any additional connection related setup or checking that is required, then override `extendConnectToModel` (don't forget to call `Super::extendConnectToModel` first in your override).  Otherwise, sockets are handled as part of the machinery of the base `Component` and you don't have to do anything to test the validity of their connections.

Reply:
Should I update BodyDragForce to fix some of these? Or shall I just focus on my volumetric contact work for this update?